Make Podman support more prominent in the documentation

rgaiacs opened the pull request:
By

- using "container" instead of Docker when possible
- mention Podman configuration

rgaiacs committed:
Use "container" instead of Docker when possible in index.md

rgaiacs committed:
Use "container" instead of Docker when possible in start.md

and add note about DOCKER_HOST environment variable.

rgaiacs committed:
Add support to DockerEngine

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

minrk commented:
My understanding was that podman was supported as long as it the engine backed by the docker cli (e.g. I use colima, not docker engine), but podman cli is _not_ supported first-class in this package. Is that accurate, @manics?

I agree generally that pointing most language to OCI terms is good, but I don't know what commitment we are making in terms of cli compatibility assumptions.

manics commented:
Originally repo2docker should've worked fine with Podman when run as a service since podman has a docker compatible API:
```
podman system service --time 0
export DOCKER_HOST=unix://$XDG_RUNTIME_DIR/podman/podman.sock
```
After https://github.com/jupyterhub/repo2docker/pull/1402 we're now using the Docker CLI. In general the Podman CLI should be compatible, but I haven't looked at how good `buildx` support is. If it currently works I've no objection to supporting if, and ensuring that any future changes to repo2docker only use buildx features that podman supports.

minrk commented:
If we've got tests running that verify it (to a reasonable degree), I'm happy to make the support official.

rgaiacs committed:
Auto configure container_cli property based on DOCKER_HOST

rgaiacs committed:
Add podman as container cli for GitHub Actions

rgaiacs committed:
Select correct hostname for Podman

rgaiacs commented:
After

- change the code to do base the logic in `DOCKER_HOST`
- add `container_cli` as a GitHub Actions matrix dimension

most of the tests using Podman fails. One of the tests fails in https://github.com/jupyterhub/repo2docker/blob/358888c6a93839dd4c60859737d886e49a4a1247/repo2docker/app.py#L633-L634 When I try to investigate it locally, I observe

```
>>> import docker
>>> client = docker.APIClient(base_url="unix:///run/user/1000/podman/podman.sock")
>>> image = client.inspect_image("localhost/r2d-tests-dockerfile-2fsimple-1759752459")
>>> image.config
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
    import platform
    ^^^^^^^^^^^^
AttributeError: 'dict' object has no attribute 'config'
>>> dir(image)
['__class__', '__class_getitem__', '__contains__', '__delattr__', '__delitem__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getitem__', '__getstate__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__ior__', '__iter__', '__le__', '__len__', '__lt__', '__ne__', '__new__', '__or__', '__reduce__', '__reduce_ex__', '__repr__', '__reversed__', '__ror__', '__setattr__', '__setitem__', '__sizeof__', '__str__', '__subclasshook__', 'clear', 'copy', 'fromkeys', 'get', 'items', 'keys', 'pop', 'popitem', 'setdefault', 'update', 'values']
>>> image.keys()
dict_keys(['Id', 'RepoTags', 'RepoDigests', 'Parent', 'Comment', 'Created', 'ContainerConfig', 'DockerVersion', 'Author', 'Config', 'Architecture', 'Os', 'Size', 'VirtualSize', 'GraphDriver', 'RootFS', 'Metadata', 'Container'])
>>> image["Config"]
{'Hostname': '', 'Domainname': '', 'User': '1000', 'AttachStdin': False, 'AttachStdout': False, 'AttachStderr': False, 'Tty': False, 'OpenStdin': False, 'StdinOnce': False, 'Env': ['PATH=/usr/local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin', 'LANG=C.UTF-8', 'GPG_KEY=A035C8C19219BA821ECEA86B64E628F8D684696D', 'PYTHON_VERSION=3.10.18', 'PYTHON_SHA256=ae665bc678abd9ab6a6e1573d2481625a53719bc517e9a634ed2b9fefae3817f', 'HOME=/tmp'], 'Cmd': ['/bin/sh', '-c', '"/bin/sh"'], 'Image': '', 'Volumes': None, 'WorkingDir': '', 'Entrypoint': None, 'OnBuild': None, 'Labels': {'io.buildah.version': '1.40.0', 'repo2docker.ref': 'None', 'repo2docker.repo': 'local', 'repo2docker.version': '2025.08.0+61.gced04a5'}}
>>> image["Config"]["WorkingDir"]
''
```

@manics do you know what might be happening?

rgaiacs committed:
Add note about Podman service to documentation

rgaiacs committed:
Enable Podman socket during test

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

rgaiacs committed:
Avoid infinite loop in test

rgaiacs committed:
Hide output of "docker version" test

rgaiacs committed:
Add WorkingDir to bridge the gap with Podman

rgaiacs committed:
Fix tests/unit/test_app.py::test_extra_buildx_build_args

rgaiacs committed:
Make DOCKER_HOST more robust

rgaiacs committed:
Prefer to use attribute over internal variable

rgaiacs commented:
With commit `efe313d256ffba5626a26c748a8ce3774ad3c5b7`, when I run

```bash
pytest -s tests/unit/test_editable.py::test_editable
```

the log shows

```
[Warning] one or more build args were not consumed: [NB_USER]
```

Why does the above warning happened? Is `--user-id` not supported for `Dockerfile` configuration?

rgaiacs committed:
Minor improve to docker.py wrap

rgaiacs committed:
Expose DOCKER_CLI from docker.py

rgaiacs committed:
Mark some tests with @pytest.mark.skipif

because Podman does not support bind mount.

Related to https://github.com/jupyterhub/repo2docker/issues/1483

rgaiacs committed:
Skip registry test for Podman

manics commented on docs/source/start.md:
````diff
@@ -2,47 +2,72 @@
-Install [Docker](https://www.docker.com), as it is required to build Docker images.
-The [Community Edition](https://docs.docker.com/install/) is available for free.
+```{important}
+Only the [Docker Engine](https://docs.docker.com/engine/) is an open source. [Docker Desktop](https://docs.docker.com/get-started/get-docker/) requires a license.
````
```suggestion
Only the [Docker Engine](https://docs.docker.com/engine/) is open source. [Docker Desktop](https://docs.docker.com/get-started/get-docker/) requires a license.
```

manics commented on docs/source/start.md:
````diff
@@ -2,47 +2,72 @@
+After complete the installation of Podman,
 
-```
+1. creates a [listening service for Podman](https://docs.podman.io/en/latest/markdown/podman-system-service.1.html) by running
````
```suggestion
1. create a [listening service for Podman](https://docs.podman.io/en/latest/markdown/podman-system-service.1.html) by running
```

manics commented on .github/workflows/test.yml:
```diff
@@ -53,6 +53,9 @@ jobs:
         python_version: ["3.13"]
+        container_cli:
+          - docker
+          - podman
```
```suggestion
```
This doubles the number of tests in the matrix, which means PR runs will take even longer, and the chance of a failure due to flakiness doubles.

I don't think we need to run all tests with Podman, just a subset, which we can do by adding them to the `include:` section below

rgaiacs committed:
Fix English mistakes in documentation

Thanks to @maniacs for the review.

rgaiacs committed:
Reduce the number of tests for Podman

rgaiacs commented:
The way that the tests are organised with [`pytest_collect_file`](https://github.com/jupyterhub/repo2docker/blob/443b85127d2f745090dd2354cb77daae0c41bcae/tests/conftest.py#L35-L39) and [`make_test_func`](https://github.com/jupyterhub/repo2docker/blob/443b85127d2f745090dd2354cb77daae0c41bcae/tests/conftest.py#L42-L94) makes hard to use pytest's built-in [`mark`](https://docs.pytest.org/en/stable/example/markers.html#marking-test-functions-and-selecting-them-for-a-run).

GitHub Actions has [`include`](https://docs.github.com/en/actions/how-tos/write-workflows/choose-what-workflows-do/run-job-variations#expanding-or-adding-matrix-configurations) and [`exclude`](https://docs.github.com/en/actions/how-tos/write-workflows/choose-what-workflows-do/run-job-variations#excluding-matrix-configurations). I think the `include` hard to understand and the `exclude` easier. Although more verbose, I'm amending this PR with `exclude`.

rgaiacs commented:
@manics and @yuvipanda are you OK for this to be merged?

minrk commented on .github/workflows/test.yml:
```diff
@@ -73,6 +76,27 @@ jobs:
             python_version: "3.10"
             repo_type: venv
+        exclude:
+          # Run all tests for both Docker and Podman is expensive.
```
Since we are excluding most cases, I think this can be replaced by a single (or few) entry under `include` for the podman case, instead of excluding most things.

rgaiacs committed:
Add comment explaining DOCKER_HOST

rgaiacs committed:
Guard subprocess.run() with try block

Co-authored-by: Min RK <benjaminrk@gmail.com>

rgaiacs committed:
Re-write matrix strategy to use include

Related to 422c0091e252b3f691b22bc3c7b1d03bee94a768

rgaiacs committed:
Remove if-block already cover in try-block

rgaiacs committed:
Fix include for GItHub Actions

manics commented:
Strangely the failing test is runingn out of disk space
`INFO     repo2docker:app.py:824 Error: committing container for step {Env:[NB_USER=runner NB_UID=1001 REPO_DIR=/srv/repo PATH=/home/runner/.local/bin:/srv/repo/.local/bin:/srv/conda/envs/notebook/bin:/srv/conda/bin:/srv/npm/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin DEBIAN_FRONTEND=noninteractive LC_ALL=en_US.UTF-8 LANG=en_US.UTF-8 LANGUAGE=en_US.UTF-8 SHELL=/bin/bash USER=runner HOME=/home/runner APP_BASE=/srv CONDA_DIR=/srv/conda NB_PYTHON_PREFIX=/srv/conda/envs/notebook NPM_DIR=/srv/npm NPM_CONFIG_GLOBALCONFIG=/srv/npm/npmrc NB_ENVIRONMENT_FILE=/tmp/env/environment.lock MAMBA_ROOT_PREFIX=/srv/conda MAMBA_EXE=/srv/conda/bin/mamba CONDA_PLATFORM=linux-64 KERNEL_PYTHON_PREFIX=/srv/conda/envs/notebook CONDA_DEFAULT_ENV=/srv/conda/envs/notebook] Command:run Args:[TIMEFORMAT='time: %3R' bash -c 'time ${MAMBA_EXE} env update -p ${NB_PYTHON_PREFIX} --file "environment.yml" && time ${MAMBA_EXE} clean --all -f -y && ${MAMBA_EXE} list -p ${NB_PYTHON_PREFIX} '] Flags:[] Attrs:map[] Message:RUN TIMEFORMAT='time: %3R' bash -c 'time ${MAMBA_EXE} env update -p ${NB_PYTHON_PREFIX} --file "environment.yml" && time ${MAMBA_EXE} clean --all -f -y && ${MAMBA_EXE} list -p ${NB_PYTHON_PREFIX} ' Original:RUN TIMEFORMAT='time: %3R' bash -c 'time ${MAMBA_EXE} env update -p ${NB_PYTHON_PREFIX} --file "environment.yml" && time ${MAMBA_EXE} clean --all -f -y && ${MAMBA_EXE} list -p ${NB_PYTHON_PREFIX} '}: copying layers and metadata for container "c60d8adc11015838b3f6e819afdad9f6054ce954adf3118bb76ca8441f83c073": writing blob: adding layer with blob "sha256:c4b8ce47904a0e175bde526208956231b1dfb32611a0a98552d58c3b96036732": processing tar file(write /srv/conda/envs/notebook/libexec/gcc/x86_64-conda-linux-gnu/9.5.0/cc1obj: no space left on device): exit status 1`

rgaiacs commented:
> Strangely the failing test is runingn out of disk space

Is a suitable solution to call [`container.remove()`](https://docker-py.readthedocs.io/en/stable/containers.html#docker.models.containers.Container.remove) and [`image.remove()`](https://docker-py.readthedocs.io/en/stable/images.html#docker.models.images.ImageCollection.remove)? Maybe after https://github.com/jupyterhub/repo2docker/blob/ca21e1b6682ad5517446483c2787d24a8724d237/tests/conftest.py#L92

rgaiacs commented:
> I'm surprised it's a problem in the first place.

I'm also surprising as the same test for Docker does not have the storage limit problem.

rgaiacs committed:
Delete container and image after test

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

rgaiacs committed:
Disable test of conda for podman

rgaiacs commented:
@manics and @minrk I disabled the test of conda for podman as mentioned in the previous message. Now, all the tests pass. This includes the `base`, `dockerfile` and `unit` test with podman.

I would like to merge this on Monday, 19 January as this pull request is already big enough and was reviewed multiple times.

dylex commented:
I'm trying to use this through binderhub with image quay.io/jupyterhub/repo2docker:2026.4.0, but it fails with `Error during build: The podman commandline client must be installed`. Does there need to be a different image that includes podman? (The rest of my podman setup worked fine with the 2024.07.0 image.)
